_source/_posts/2024-01-24-spring-boot-authorization.md

```diff
@@ -0,0 +1,569 @@
+
+{% img blog/spring-boot-authorization/whatabyte-ui.png alt:"WHATABYTE Dashboard" width:"800" %}{: .center-image }
+
+On the top right corner, click **Add Item**, and a pre-populated form will display. Click on **Save** and verify the user request is authorized in the server.
+
+## Role-Based Access Control (RBAC)
+
+If you disable the authentication features in the WHATABYTE client, it will allow you to attempt CRUD operations from the UI, but your API server will reject item modifications as an unauthenticated user.
+
+{% img blog/spring-boot-authorization/whatabyte-unauth.png alt:"WHATABYTE Dashboard Unauthenticated" width:"800" %}{: .center-image }
+
+Besides requiring a user to be authenticated, it is a standard practice to create roles for authorizing write operations.
+
+### Create and assign roles
+
+In the WHATABYTE client settings, re-enable the authentication features, and also enable RBAC. Set `menu-admin` in the _User Role_ text box. Click on **Save**.
+
+Now if you sign in with the user you created, the UI will not display the links to perform write operations, as the role has not yet been assigned.
+
+First, the role must be defined in the Auth0 tenant as well. You can use the following Auth0 CLI command:
+
+```shell
+auth0 roles create
+```
+
+Assign the role to the user you created:
+
+```shell
+auth0 users roles assign
+```
+
+> NOTE: You can find the user id with `auth0 users search --query email:<email>`
+
+Follow the steps, you will see the output below:
+
+```text
+=== dev-avup2laz.us.auth0.com user roles (1)
+
+  ID                    NAME           DESCRIPTION                                
+  rol_Nf2xmon4GzTbLU9T  menu-admin     Admin role for menu item write operations  
+```
+
+### Mapping the roles to token claims
+
+The role `menu-admin` and its permissions must be mapped to a claim in the accessToken, to make them available in the API for authorization. With [Auth0 Actions](https://auth0.com/docs/customize/actions) you can customize the Login flow to map the user roles to a custom claim.
+
+First [configure your preferred editor](https://github.com/auth0/auth0-cli#customization) to use with the Auth0 CLI:
+
+```shell
+export EDITOR=nano
+```
+
+Then create the Login Action:
+
+```shell
+auth0 actions create
+```
+
+Set the name __Add Roles__, and select **post-login** for the Trigger. When the editor opens, set the following implementation for the `onExecutePostLogin` function.
+
+```javascript
+exports.onExecutePostLogin = async (event, api) => {
+  const namespace = 'https://menu-api.okta.com';
+  if (event.authorization) {
+    api.idToken.setCustomClaim('preferred_username', event.user.email);
+    api.idToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
+    api.accessToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
+  }
+}
+```
+
+Save the file. You should see the output below once the action is created.
+
+```text
```

Comment:
This long output seems unncessary, can we just say `You should see the output once the action is created`

Reply:
Removed.